app/controllers/application_controller.rb

```diff
@@ -5,6 +5,158 @@ class ApplicationController < Sinatra::Base
+
+    enable :sessions
+    set :session_secret, "secret_fwitter"
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect '/tweets'
+    elsif !logged_in?
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if logged_in?
+      redirect '/tweets'
+    end
+    if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect '/signup'
+    else
+      @user = User.create(:username => params[:username], :password => params[:password])
+      @user.save
+      session[:user_id] = @user.id
+
+      redirect '/tweets'
+    end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  get '/failure' do
+   erb :'users/failure'
+ end
+
+  get '/logout' do
+    if logged_in?
+      session.clear
+      redirect '/login'
+    else
+      redirect '/tweets'
+    end
+  end
+
+  get '/tweets' do
+    if logged_in?
+      @user = User.find(session[:user_id])
+      @tweets = Tweet.all
+      erb :'/tweets/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/new' do
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do
+    if !params[:content].empty?
+      @tweet = Tweet.create(content: params[:content], user_id: current_user.id)
+      redirect to "/tweets/#{@tweet.id}"
+    else
+      redirect to '/tweets/new'
+    end
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      erb :'/tweets/edit_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      @tweet.update(content: params[:content])
+      if !params[:content].empty?
+        @tweet.save
+        redirect "/tweets/#{@tweet.id}"
+      else
+        redirect "/tweets/#{@tweet.id}/edit"
+      end
+    else
+      redirect '/login'
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet && current_user.id == @tweet.user_id
+        @tweet.delete
+        redirect '/tweets'
+      end 
+    elsif !logged_in?
+      redirect "/login"
+    else
+      redirect "/tweet/#{@tweet.id}"
+    end
+  end
+
+
+  helpers do
+    def logged_in?
+      !!session[:user_id]
+    end
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```